fix(ast): fix `ContentEq` and `ContentHash` impls for literal types

overlookmotel opened the pull request:
Fix implementations of `ContentEq` and `ContentHash` for literal AST types:

* `NullLiteral::content_hash` is a no-op, same as other types which only contain a `Span`.
* `NumericLiteral::content_eq` and `content_hash` ignore `base` field.
* `NumericLiteral::content_hash` works around `0.0 == -0.0`.
* `BigIntLiteral::content_eq` and `content_hash` ignore `base` field.
* `StringLiteral::content_hash` ignore `raw` field.
* `RegExpLiteral::content_eq` and `content_hash` consider 2 `RegExp`s to be equal if they are printed the same (regardless of whether they were parsed by `oxc_regular_expression` or not).

Additionally, implement `StringLiteral::content_eq` manually to avoid "special case" logic in `oxc_ast_tools`.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8426?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#8427** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8427?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8426** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8426?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8426?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8425** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8425?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8426).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
@Boshen would you mind taking a look at this? Sounds like you're using `ContentEq` in minifier.

I think it's all uncontroversial except:

#### NaN in `NumericLiteral`

Is it OK that `f64::NAN != f64::NAN`, and therefore 2 x `NumericLiteral`s containing `NaN` will NOT be `content_eq` (same as they're not `partial_eq`)? We do now use `f64::NAN` in the codebase, so I'm wondering if this may cause a problem?

This PR doesn't change that behavior. I just hadn't noticed this potential problem before.

#### Ignore `base` field for `NumericLiteral` and `BigIntLiteral`

I think `base` field is not useful for comparisons, in the same way as `raw` field isn't. It relates to how the value appears in source, but is otherwise irrelevant. `0x8` and `8` are the same number.

Side note: Should we get rid of `ContentHash` entirely? Or put it behind a feature? I don't think we use it anywhere. If I remember right, original use case was for the linter to compare if 2 switch cases are the same, but I think we use `ContentEq` for that now.

Boshen commented:
### Merge activity

* **Jan 10, 9:41 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
fix(ast): fix `ContentEq` and `ContentHash` impls for literal types (#8426)

Fix implementations of `ContentEq` and `ContentHash` for literal AST types:

* `NullLiteral::content_hash` is a no-op, same as other types which only contain a `Span`.
* `NumericLiteral::content_eq` and `content_hash` ignore `base` field.
* `NumericLiteral::content_hash` works around `0.0 == -0.0`.
* `BigIntLiteral::content_eq` and `content_hash` ignore `base` field.
* `StringLiteral::content_hash` ignore `raw` field.
* `RegExpLiteral::content_eq` and `content_hash` consider 2 `RegExp`s to be equal if they are printed the same (regardless of whether they were parsed by `oxc_regular_expression` or not).

Additionally, implement `StringLiteral::content_eq` manually to avoid "special case" logic in `oxc_ast_tools`.

Boshen commented:
> so I'm wondering if this may cause a problem?

Not a huge deal if we don't minify a piece of code that contain literals `NaN`s ... such code may not exist anyway.
